src/mmap_unix.rs

```diff
@@ -154,25 +158,103 @@ impl MmapRegion {
             return Err(Error::Mmap(io::Error::last_os_error()));
         }
 
-        Ok(Self {
-            addr: addr as *mut u8,
-            size,
+        let mut mmap_region = Self {
+            start_addr: addr,
+            total_size: size,
+
+            data_addr: addr,
+            data_size: size,
+
             file_offset,
             prot,
             flags,
-        })
+        };
+
+        // Try to surround the `MmapRegion` with guard pages.
+        // If it's not possible, return it as it is.
+        mmap_region.try_guard();
+
+        Ok(mmap_region)
+    }
+
+    /// Tries to surround the `MmapRegion` with guard pages.
+    ///
+    /// +-----+-------------+---------------+-----+-----------------------+-------------+-----+
+    /// | ... | memory page |  memory page  | ... |      memory page      | memory page | ... |
+    /// +-----+-------------+---------------+-----+-------------+---------+-------------+-----+
+    /// | ... |  guard page | mapping start | ... | mapping end | padding |  guard page | ... |
+    /// +-----+-------------+---------------+-----+-------------+---------+-------------+-----+
+    ///
+    ///  # Return
+    ///
+    /// Returns a boolean indicating whether the operation succeeded.
+    ///
+    pub fn try_guard(&mut self) -> bool {
+        // Check if the `MmapRegion` is already guarded
+        if self.start_addr != self.data_addr && self.total_size != self.data_size {
+            return false;
+        }
+
+        let page_size = match unsafe { libc::sysconf(libc::_SC_PAGESIZE) } {
```

Comment:
Should we align the redzone to hugepage size(2M/1G)?
Otherwise it may conflicts with THP.